fix(smolagents): support smolagents nested llm input message contents

andstor opened the pull request:
Adds suport for smolagents' new nested "text" llm input message contents.
Fixes #1237

axiomofjoy commented:
@andstor Rebased to incorporate your other merged PRs and fixed types. The tests appear to be passing on `main`. Can you help me understand where this change is needed, e.g., with a code snippet that fails to instrument input message attributes on `main`?

andstor commented:
> @andstor Rebased to incorporate your other merged PRs and fixed types. The tests appear to be passing on `main`. Can you help me understand where this change is needed, e.g., with a code snippet that fails to instrument input message attributes on `main`?

@axiomofjoy The tests were passing on `main` because I had added backward compatibility to allow simple string content. Since this is no longer used at all in smolagents, I have removed it. This will trigger tests to fail, so I have updated them accordingly. The cassettes should likely be updated as well for good measure, but I was unable to run this successfully.

axiomofjoy commented:
> > @andstor Rebased to incorporate your other merged PRs and fixed types. The tests appear to be passing on `main`. Can you help me understand where this change is needed, e.g., with a code snippet that fails to instrument input message attributes on `main`?
> 
> @axiomofjoy The tests were passing on `main` because I had added backward compatibility to allow simple string content. Since this is no longer used at all in smolagents, I have removed it. This will trigger tests to fail, so I have updated them accordingly. The cassettes should likely be updated as well for good measure, but I was unable to run this successfully.

Thanks so much @andstor! We'll get this reviewed and merged.

nate-mar commented on python/instrumentation/openinference-instrumentation-smolagents/src/openinference/instrumentation/smolagents/_wrappers.py:
```diff
@@ -164,23 +164,23 @@ def __call__(
+        yield from process_message(0, "user", prompt)
     elif isinstance(messages := arguments.get("messages"), list):
-        for i, message in enumerate(messages):
+        i = 0
+        for message in messages:
```
Can we still use `enumerate` here instead of `I` to handle the incrementing?

andstor replied:
Yes, enumerate works fine. This was simply a leftover.

nate-mar commented on python/instrumentation/openinference-instrumentation-smolagents/src/openinference/instrumentation/smolagents/_wrappers.py:
```diff
@@ -164,23 +164,23 @@ def __call__(
+            role, content = message.get("role"), message.get("content")
+            if isinstance(content, list) and role:
+                for subcontent in content:
+                    if isinstance(subcontent, dict) and (text := subcontent.get("text")):
```
Is there support for non text messages, like audio/image?

nate-mar reviewed (commented):
A few small comments --  I'm happy to make the enumerate change myself (unless I'm missing something here) and merge, if you'd like! Thanks again for the fix here! Really appreciate it!
